From 2745372ff05b59454e617f148076c3eab1121eef Mon Sep 17 00:00:00 2001
From: Zhibin Liu <zhibinliu@google.com>
Date: Wed, 21 May 2025 23:57:59 -0700
Subject: [PATCH] Check package name format before using it for SPA nav for udc

Bug: b/409780975
Test: unit test
Flag: EXEMPT bug fix
(cherry picked from https://googleplex-android-review.googlesource.com/q/commit:e806e18728782060c3d151fc515a503bcdd71209)
Merged-In: Ife615c78c48a2d6f5b938c880c9c2128c5551644
Change-Id: I3aba7bcc34b5ecf3d35d06e531ea01ab908b200f
---
 src/com/android/settings/spa/SpaActivity.kt         |  2 +-
 .../src/com/android/settings/spa/SpaActivityTest.kt | 13 +++++++++++++
 2 files changed, 14 insertions(+), 1 deletion(-)

diff --git a/src/com/android/settings/spa/SpaActivity.kt b/src/com/android/settings/spa/SpaActivity.kt
index 2b52b21af0..6b7a3668b7 100644
--- a/src/com/android/settings/spa/SpaActivity.kt
+++ b/src/com/android/settings/spa/SpaActivity.kt
@@ -64,7 +64,7 @@ class SpaActivity : BrowseActivity() {
 
         @JvmStatic
         fun Context.startSpaActivityForApp(destinationPrefix: String, intent: Intent): Boolean {
-            val packageName = intent.data?.schemeSpecificPart ?: return false
+            val packageName = intent.data?.schemeSpecificPart?.takeIf { Regex("^([a-zA-Z]\\w*\\.)*[a-zA-Z]\\w*$").matches(it) } ?: return false
             startSpaActivity("$destinationPrefix/$packageName/${UserHandle.myUserId()}")
             return true
         }
diff --git a/tests/spa_unit/src/com/android/settings/spa/SpaActivityTest.kt b/tests/spa_unit/src/com/android/settings/spa/SpaActivityTest.kt
index 1b2a7b1ee9..1aa8d17eb8 100644
--- a/tests/spa_unit/src/com/android/settings/spa/SpaActivityTest.kt
+++ b/tests/spa_unit/src/com/android/settings/spa/SpaActivityTest.kt
@@ -36,6 +36,8 @@ import org.junit.Test
 import org.junit.runner.RunWith
 import org.mockito.ArgumentCaptor
 import org.mockito.Mock
+import org.mockito.Mockito.any
+import org.mockito.Mockito.never
 import org.mockito.Mockito.verify
 import org.mockito.MockitoSession
 import org.mockito.quality.Strictness
@@ -115,6 +117,17 @@ class SpaActivityTest {
             .isEqualTo("Destination/package.name/${UserHandle.myUserId()}")
     }
 
+    @Test
+    fun startSpaActivityForApp_hasMalformedPackageName() {
+        val intent = Intent().apply {
+            data = Uri.parse("package:package.name/10#")
+        }
+
+        context.startSpaActivityForApp(DESTINATION, intent)
+
+        verify(context, never()).startActivity(any())
+    }
+
     private companion object {
         const val DESTINATION = "Destination"
         const val PACKAGE_NAME = "package.name"
-- 
2.50.1.351.gcf940e82a1-goog

